ui/src/main/java/org/jtaccuino/core/ui/completion/CompletionSelectionModel.java

```diff
@@ -0,0 +1,224 @@
+ */
+package org.jtaccuino.core.ui.completion;
+
+import java.util.Objects;
+import javafx.collections.FXCollections;
+import javafx.collections.ListChangeListener;
+import javafx.collections.ObservableList;
+import javafx.scene.control.MultipleSelectionModel;
+
+class CompletionSelectionModel extends MultipleSelectionModel<CompletionItem> {
+
+    private final ObservableList<CompletionItem> items;
+    private final ObservableList<Integer> selectedIndicesList = FXCollections.observableArrayList();
+    private final ObservableList<CompletionItem> selectedItemsList = FXCollections.observableArrayList();
+
+    public CompletionSelectionModel(ObservableList<CompletionItem> items) {
+        this.items = items;
+        setSelectedIndex(-1);
+        setSelectedItem(null);
+        items.addListener((ListChangeListener<CompletionItem>) c -> {
+            while (c.next()) {
+                if (c.wasAdded() || c.wasRemoved() || c.wasPermutated()) {
+                    int currentSelected = getSelectedIndex();
+                    if (currentSelected != -1 && (currentSelected >= items.size() || items.get(currentSelected) == null)) {
+                        int newSelection = findNextSelectableIndex(currentSelected, false); // Try forward from current index
+                        if (newSelection == -1) {
+                            newSelection = findPrevSelectableIndex(currentSelected - 1, false);
+                        }
+                        if (newSelection != -1) {
+                            select(newSelection);
+                        } else {
+                            clearSelection();
+                        }
+                    }
+                }
+            }
+        });
+    } // Try forward from current index
+
+    @Override
+    public ObservableList<Integer> getSelectedIndices() {
+        return selectedIndicesList;
+    }
+
+    @Override
+    public ObservableList<CompletionItem> getSelectedItems() {
+        return selectedItemsList;
+    }
+
+    @Override
+    public void select(int index) {
+        clearAndSelect(index);
+    }
+
+    @Override
+    public void select(CompletionItem obj) {
+        int index = items.indexOf(obj);
+        if (index != -1 && obj != null) {
+            select(index);
+        } else {
+            clearSelection();
+        }
+    }
+
+    @Override
+    public void selectNext() {
+        int currentIndex = getSelectedIndex();
+        int nextIndex = findNextSelectableIndex(currentIndex + 1, false);
+        select(nextIndex);
+    }
+
+    @Override
+    public void selectPrevious() {
+        int currentIndex = getSelectedIndex();
+        int prevIndex = findPrevSelectableIndex(currentIndex - 1, false);
+        select(prevIndex);
+    }
+
+    private int findNextSelectableIndex(int startIndex, boolean wrapAround) {
+        int maxIndex = items.size();
+        for (int i = startIndex; i < maxIndex; i++) {
+            if (i >= 0 && i < items.size()) {
```

Comment:
You can avoid the inner if by incorporating its conditions into the loop
```suggestion
        for (int i = Math.max(0, startIndex); i < Math.min(itmes.size(), maxIndex); i++) {
```